Updates for Terraform policy docs (#2864)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/write/examples.mdx

## Encrypt EBS volumes

This example demonstrates a basic resource policy that ensures all AWS EBS volumes are encrypted and use an approved KMS key. It also shows how to use `core::getdatasource()` to retrieve data sources and the `data` test block to mock them.

### Policy

> The policy checks that each `aws_ebs_volume` resource is encrypted and uses the approved KMS key. It uses two separate policy blocks: one to enforce encryption, and another (with a filter) to enforce the KMS key requirement only on encrypted volumes.

Reviewer: ```suggestion
The policy checks that each `aws_ebs_volume` resource is encrypted and uses the approved KMS key. It applies a policy block that enforces encryption. A second block uses a filter to enforce the KMS key requirement on encrypted volumes.
```

<CodeBlockConfig filename="policies/encrypted_ebs.policy.hcl">

```hcl
locals {
  approved_kms_key = core::getdatasource("aws_kms_key", {
    key_id = "alias/approved-ebs-key"
  })
}

resource_policy "aws_ebs_volume" "ebs_encrypted" {
  enforce {
    condition     = attrs.encrypted == true
    error_message = "EBS volume is not encrypted"
  }
}

resource_policy "aws_ebs_volume" "approved_kms_key" {
  filter = attrs.encrypted == true

  enforce {
    condition     = attrs.kms_key_id == local.approved_kms_key.id
    error_message = "EBS volume must use the approved KMS key: ${local.approved_kms_key.id}"
  }
```

</CodeBlockConfig>

### Tests

> This test demonstrates using the `data` block to mock a KMS key data source. It includes three test cases: one that passes both policies, one that fails because it's not encrypted, and one that fails because it uses the wrong KMS key.

Reviewer: ```suggestion
This test demonstrates using the `data` block to mock a KMS key data source. It includes a test case that passes both policies, a case that fails when the data source isn't encrypted, and a case that fails when the data source uses the wrong KMS key.
```
Right? I wasn't sure what "it's" is referring to. Also, don't use a colon mid sentence per the style guide.

## Encrypt EBS volumes

This example demonstrates a resource policy that ensures all AWS EBS volumes are encrypted and use an approved KMS key. It also shows how to use `core::getdatasource()` to retrieve data sources and the `data` test block to mock them.

### Policy

The policy checks that each `aws_ebs_volume` resource is encrypted and uses the approved KMS key. It applies a policy block that enforces encryption. A second block uses a filter to enforce the KMS key requirement on encrypted volumes.

<CodeBlockConfig filename="policies/encrypted_ebs.policy.hcl">

```hcl
locals {
  approved_kms_key = core::getdatasource("aws_kms_key", {
    key_id = "alias/approved-ebs-key"
  })
}

resource_policy "aws_ebs_volume" "ebs_encrypted" {
  enforce {
    condition     = attrs.encrypted == true
    error_message = "EBS volume is not encrypted"
  }
}

resource_policy "aws_ebs_volume" "approved_kms_key" {
  filter = attrs.encrypted == true

  enforce {
    condition     = attrs.kms_key_id == local.approved_kms_key.id
    error_message = "EBS volume must use the approved KMS key: ${local.approved_kms_key.id}"
  }
}
```

</CodeBlockConfig>

### Tests

> This test demonstrates using the `data` block to mock a KMS key data source. It includes a test case that passes both policies, a case that fails when the data source isn't encrypted, and a case that fails when the data source uses the wrong KMS key.

Reviewer 2: "a case that fails when the data source isn't encrypted" -- should be "when the volume isn't encrypted".

}

resource "aws_kms_key" "ebs" {
  description             = "KMS key for EBS volume encryption"
  deletion_window_in_days = 10
}

resource "aws_kms_alias" "ebs" {
  name          = "alias/approved-ebs-key"
  target_key_id = aws_kms_key.ebs.key_id
}

data "aws_kms_key" "approved" {
  key_id = aws_kms_alias.ebs.name
}

resource "aws_ebs_volume" "pass" {
  availability_zone = "us-west-2a"
  size              = 10
  encrypted         = true
  kms_key_id        = data.aws_kms_key.approved.id
}

resource "aws_ebs_volume" "fail_not_encrypted" {
  availability_zone = "us-west-2b"
  size              = 10
  encrypted         = false
}

resource "aws_ebs_volume" "fail_wrong_key" {
  availability_zone = "us-west-2c"
  size              = 10
  encrypted         = true
}
```

</CodeBlockConfig>

## Ensure Azure managed disks have encryption settings

This example demonstrates a resource policy that validates nested attributes in Azure managed disks. It also shows how to use `input` blocks to parameterize policies, the `inputs` block in tests to set and override input values, and the `mandatory_overridable` enforcement level to allow exceptions with approval.

### Policy

> The policy uses an input variable to make the encryption requirement configurable and sets `enforcement_level = "mandatory_overridable"` to allow users with appropriate permissions to override the policy when needed (e.g., for development environments).

Reviewer: ```suggestion
The policy uses an input variable to make the encryption requirement configurable. It sets `enforcement_level = "mandatory_overridable"` to allow users with appropriate permissions to override the policy when needed, such as for development environments.
```

## Verify module sources and versions

This example demonstrates a module policy that validates module sources and versions using meta-attributes and the semverconstraint function. It uses `enforcement_level = "advisory"` to provide recommendations rather than hard requirements, allowing teams flexibility when needed.

### Policy

> The policy recommends that VPC modules are sourced from an internal private registry and use version 2.0.0 or higher. The advisory enforcement level means violations will generate warnings but won't block operations.

Reviewer: ```suggestion
The policy recommends that VPC modules are sourced from an internal private registry and use version 2.0.0 or higher. Because the enforcement level is set to `advisory`, violations generate warnings but don't block operations.
```

## Validate CloudTrail S3 bucket ACL

> This example demonstrates using `core::getresources()` to validate relationships between resources. It uses `enforcement_level = "mandatory_overridable"` to allow exceptions for special cases (e.g., public audit trails) with appropriate approval.

Reviewer: ```suggestion
This example demonstrates using `core::getresources()` to validate relationships between resources. It uses `enforcement_level = "mandatory_overridable"` to allow exceptions for special cases, such as public audit trails, with appropriate approval.
```

### Policy

> The policy ensures that CloudTrail resources are associated with S3 buckets that have private ACLs. The mandatory_overridable enforcement level allows users with appropriate permissions to override this requirement when necessary.

Reviewer: ```suggestion
The policy ensures that CloudTrail resources are associated with S3 buckets that have private ACLs. Because the enforcement level is set to `mandatory_overridable`, users with appropriate permissions can override this requirement when necessary.
```

```

</CodeBlockConfig>

## Prevent EBS volume downsizing

> This example demonstrates using the `operations` parameter to control when policies are enforced and the `prior_attrs` attribute to access previous resource state. This is useful for preventing dangerous operations like reducing volume sizes that could cause data loss.

Reviewer: ```suggestion
This example demonstrates using the `operations` parameter to control when policies are enforced and the `prior_attrs` attribute to access previous resource state. This is useful for preventing dangerous operations such as reducing volume sizes that could cause data loss.
```

### Policy

> The policy uses `operations = ["update"]` to only enforce during update operations (not create or delete) and compares `prior_attrs.size` with `attrs.size` to prevent downsizing.

Reviewer: ```suggestion
The policy uses `operations = ["update"]` so that only enforce during update operations trigger evaluation. During evaluation, the policy compares `prior_attrs.size` with `attrs.size` to prevent downsizing.
```

<CodeBlockConfig filename="policies/prevent_downsize.policy.hcl">

```hcl
resource_policy "aws_ebs_volume" "prevent_downsize" {
  operations = ["update"]

  enforce {
    condition     = attrs.size >= prior_attrs.size
    error_message = "EBS volume size cannot be reduced from ${prior_attrs.size}GB to ${attrs.size}GB. Downsizing volumes can cause data loss."
  }
}
```

</CodeBlockConfig>

### Tests

> This test demonstrates using `prior_attrs` in test resources to simulate update operations. It includes three scenarios: increasing size (pass), keeping same size (pass), and decreasing size (fail).

Reviewer: ```suggestion
This test demonstrates using `prior_attrs` in test resources to simulate update operations. It includes a scenario that passes when the EBS volume increases in size, a scenario that passes when the volume remains the same size, and a scenario that fails when the volume size decreases.
```

```

</CodeBlockConfig>

## Prevent EBS volume downsizing

This example demonstrates using the `operations` parameter to control when policies are enforced and the `prior_attrs` attribute to access previous resource state. This is useful for preventing dangerous operations such as reducing volume sizes that could cause data loss.

### Policy

> The policy uses `operations = ["update"]` so that only enforce during update operations trigger evaluation. During evaluation, the policy compares `prior_attrs.size` with `attrs.size` to prevent downsizing.

Reviewer 2: This sentence is garbled: "so that only enforce during update operations trigger evaluation". Suggest: "so that the policy is only enforced during update operations".

<CodeBlockConfig filename="tests/prevent_downsize.policytest.hcl">

```hcl
policytest {
  targets = ["../policies/prevent_downsize.policy.hcl"]
}

resource "aws_ebs_volume" "pass_increase" {
  prior_attrs = {
    availability_zone = "us-east-1a"
    size              = 50
    encrypted         = true
  }

  attrs = {
    availability_zone = "us-east-1a"
    size              = 100
    encrypted         = true
  }
}

resource "aws_ebs_volume" "pass_same_size" {
  prior_attrs = {
    availability_zone = "us-east-1a"
    size              = 50
    encrypted         = true
  }

  attrs = {
    availability_zone = "us-east-1a"
    size              = 50
    encrypted         = true
  }
}

resource "aws_ebs_volume" "fail_decrease" {
  expect_failure = true

  prior_attrs = {
    availability_zone = "us-east-1a"
    size              = 100
    encrypted         = true
  }

  attrs = {
    availability_zone = "us-east-1a"
    size              = 50
    encrypted         = true
  }
}
```

</CodeBlockConfig>

### Validate and test

Validate the policy syntax:

```shell-session
$ tfpolicy validate --policies=policies/prevent_downsize.policy.hcl

> Success! Policy is valid.

Reviewer 2: The CLI actually prints `Success! Terraform policies validated successfully.` (see `validate/command.go` in terraform-policy-cli), which is also what `reference/cli/validate.mdx` shows in this PR. The `Success! Policy is valid.` string appears throughout the pre-existing examples too, so this may be worth a file-wide fix.
